### ISSUE TITLE
Document IGP for Relayer operators

### ISSUE THREAD

**Issue by nambrot** [hyperlane-xyz/hyperlane-monorepo#1857]
Right now its unclear what relayer operators should know about IGPs, specifically about how to think about quoting and charging payments, when and when not to message filter, etc.

Likely blocked on #1848 and then a separate page that can be linked to from the relayer configuration section.

Probably start with no enforcement and an upgradable IGP.
Possibly need a separate guide to deploy static gas oracles as the next step that is a separate script in hyperlane-deploy.

Gas enforcement should maybe default to no enforcement

**Comment by asaj** [comment]
I would suggest that we do not try to get gas payments working for PI chains until we've sorted out more of the kinks

**Comment by nambrot** [comment]
Tend to agree, in that case we should IMO 1) default to no enforcement and 2) tell deployers that that is the case so that they are not surprised

**Comment by nambrot** [comment]
PI deployers need to understand that that currently can't use the default IGPs when sending messages to their chain, so they will need to deploy new IGP and we will eventually need to build composing IGPs

**Comment by tkporter** [comment]
> I would suggest that we do not try to get gas payments working for PI chains until we've sorted out more of the kinks

I agree - opened https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/1871

**Comment by nambrot** [comment]
Talking to DAM, I think we do need some kind of IGP though, otherwise you are forcing Router users to use a different function call. IMO we should have a "NoOpIGP" sooner rather than later

**Comment by nambrot** [comment]
I.e. idk what Router app they used, but the suggestion of not using an IGP at all was responded with "uff now I have to do another contract deployment" and agree that that is not ideal

**Comment by asaj** [comment]
> IMO we should have a "NoOpIGP" sooner rather than later

Can't we just skip the `payForGas` call if no IGP is configured?

**Comment by asaj** [comment]
So it sounds like on the PI side we want to document that the relayer will be subsidizing messages to/from the PI chain, and point them to the whitelist/blacklist docs if they want more fine grained control.

And maybe we need to modify Router.sol to skip `payForGas` if the IGP is not configured?

**Comment by nambrot** [comment]
Maybe taking a step back, I believe we have the following issue:

- Currently, we
  - Deploy a default IGP implementation on the PI chain
  - Deploy no IGPs on the remote chains
  - We don't really say much about IGPs/gas payments in PI/relayer docs
  - Router applications like Warp Routes require IGPs to be set in config/initialization
  - Router applications typically use Router#dispatchWithGas which 
  - We imply in the warp route docs to pick these addresses from the docs
  - If you use the default IGP to send to PI chain with a Router application, it will revert with `!gas oracle`
  - If you use the deployed IGP on the PI chain to send to default chain with a Router application, right now you will pay 1 wei, soon with PR merging you will also get `!gas oracle`
  - Warp route UI assumes all warp routes have IGPs
- Possible changes:
  - Adjust warp route config docs to be more clear that choosing the right IGP matters
  - Modify Router.sol to not validate that an IGP is empty (similar to ISMs) and then have dispatchWithGas ignore payments if its not set
    - IMO feels like a heavier change. Having routers always have IGPs seems like a useful invariant, or maybe one that is harder to undo across all invocations. Having to deploy IGPs (even ones that don't do anything), seems preferrable as an acknowledgement of the path you are taking
  - Create a static igp that just chages a constant amount (which can be 0) and in PI guide say to use those when sending messages to/from PI chains

Regardless of PI we should:
- have general docs on how payments work at a high level (i.e not a guide but under the protocol section)
- improve gas related configuration options in docs

Probably punt for the future:
- have PI guide show how to specify IGPs with gas oracles

**Comment by samtvlabs** [comment]
We find this useful as we are currently designing our gas operations. Would be keen to understand what changes are expected to happen to PI deployment @nambrot

**Comment by nambrot** [comment]
@asaj Is this ticket done?

**Comment by nambrot** [comment]
Though actually now that I'm looking at it, I think we still don't really document how relayers should think about gas enforcement and what their options are. We have a page about whitelisting and blacklisting, but I think that's it?

**Comment by asaj** [comment]
> Though actually now that I'm looking at it, I think we still don't really document how relayers should think about gas enforcement and what their options are. We have a page about whitelisting and blacklisting, but I think that's it?


@nambrot I'm not sure we have a great story here tbh. The longer term answer is probably something about maintaining your own gas oracles (or using ones that someone else maintains) but in the meantime...

**Comment by nambrot** [comment]
I agree that full blown documentation of the gas oracles is probably overkill right now (especially on the PI guide), though not listing options and indicating that the default policy is `GasPaymentPolicyNone` also seems wrong.

Wdyt about just on the `Configurations` page list at the bottom to specify the gaspayment policy options, maybe document the default none policy. For minimum payment and on chain enforcement options, just link to the code and indicate it is advanced?
